### PR TITLE
feat: improve Python doctor diagnostics

### LINKED ISSUE #203: Improve Python diagnostics in the doctor command
## Summary

Improve deterministic Python project discovery and diagnostics in `azure-functions-skills doctor`. Python v2 applications are not currently inventoried as completely as Node.js applications, which can prevent existing binding and storage checks from running and leaves several Python-specific deployment problems undetected.

This issue tracks the implementation in dependency order. Tier 1 checks should remain deterministic, fast, offline-capable, and conservative about false positives.

### Phase 1: project and function discovery

- [ ] Detect Python projects without requiring `requirements.txt`, using signals such as `function_app.py`, `pyproject.toml`, and `FUNCTIONS_WORKER_RUNTIME=python`.
- [ ] Define deterministic precedence when multiple language signals are present.
- [ ] Detect Python v1, v2, and mixed programming-model layouts.
- [ ] Discover Python v2 functions and trigger types from `FunctionApp` and `Blueprint` registrations, including imported aliases.
- [ ] Feed discovered Python triggers and bindings into the existing `FunctionInfo` inventory so storage, connection, and binding checks work for Python v2 projects.
- [ ] Add fixtures and regression tests proving non-HTTP Python v2 triggers no longer produce silent false negatives.

### Phase 2: dependency analysis foundation

- [ ] Add a shared Python dependency parser instead of parsing `requirements.txt` independently in each check.
- [ ] Support normalized package names, extras, environment markers, inline comments, hashes, direct URLs, editable requirements, and `-r`/`-c` includes with safe workspace-bound path handling.
- [ ] Read supported dependency manifests such as `requirements.txt` and `pyproject.toml` without reporting one as missing when the other is authoritative.
- [ ] Migrate the existing unpinned-requirements and missing-lockfile checks to the shared parser.
- [ ] Add tests for malformed manifests, recursive includes, duplicate declarations, and package-name normalization.

### Phase 3: high-confidence Python checks

- [ ] Report a missing dependency manifest when a Python Functions project has external imports but no supported manifest.
- [ ] Report a missing or unsupported `azure-functions` dependency.
- [ ] Warn when `azure-functions-worker` is explicitly declared; the worker is managed by the Functions platform and should not normally be pinned by the application.
- [ ] Detect decorated Blueprint functions that are not registered through `app.register_functions(...)`.
- [ ] Detect mixed v1/v2 layouts and assign severity only after confirming supported Azure Functions behavior.
- [ ] Include actionable file and line information where source parsing provides it.

### Phase 4: security and deployment checks

- [ ] Extend tracked-secret-file detection to cover `local.settings.json`, including files already tracked despite a later `.gitignore` entry.
- [ ] Add an informational native-dependency compatibility finding for packages commonly requiring platform-specific wheels or build tooling; do not imply that their use is inherently incorrect.
- [ ] Warn when test, virtual-environment, cache, or bytecode artifacts are likely to enter the deployment package because `.funcignore` coverage is missing or incomplete.
- [ ] Add a Durable configuration coherence check after Python v2 trigger discovery is available.
- [ ] Add only low-severity local observability guidance when Application Insights settings are absent; do not claim the deployed Azure resource is unconfigured based only on local files.

### Phase 5: deep-analysis references and documentation

- [ ] Update the canonical doctor skill references with Blueprint registration, explicit `azure-functions-worker`, and native-dependency guidance.
- [ ] Keep deterministic checks out of the deep-analysis result when Tier 1 has already reported them.
- [ ] Document the new check IDs, severities, applicability, limitations, and remediation.
- [ ] Regenerate plugin payloads from `templates/` and verify them rather than editing generated files directly.

### Phase 6: CI reporting follow-up

- [ ] Evaluate SARIF output as a separate reporting enhancement for GitHub code scanning.
- [ ] Evaluate profiles and custom rule configuration only after the deterministic Python checks and stable check IDs are complete.

## Out of scope

- Checking whether the current shell has an activated virtual environment.
- Validating the local `python` executable or Azure Functions Core Tools installation.
- Treating generic ASGI/WSGI source patterns as errors without a concrete Functions configuration conflict.
- Duplicating existing `host.json`, extension bundle, runtime-stack, or deep semantic checks.
- Autofixing application files.

## Acceptance criteria

- Python projects are detected even when `requirements.txt` is missing.
- Python v2 HTTP, timer, queue, Service Bus, Event Hubs, blob, and Durable registrations are inventoried when represented by supported SDK patterns.
- Existing binding, connection, and host-storage checks receive the Python v2 inventory.
- Dependency checks share one parser and handle supported manifest syntax without contradictory findings.
- New heuristic checks use non-blocking severity and clearly communicate uncertainty.
- Unit tests cover v1, v2, Blueprint, mixed-model, malformed-input, and dependency-manifest cases.
- Relevant CLI E2E fixtures pass in isolated workspaces.
- `npm run check` passes after canonical templates and generated payloads are synchronized.

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
## Summary

- discover Python v1, v2, mixed, Blueprint, and Durable FunctionApp functions for deterministic doctor checks
- add shared Python dependency parsing and diagnostics for manifests, worker dependencies, native packages, deployment artifacts, Durable configuration, and observability
- add scenario-specific Python bad-app fixtures, canonical doctor skill guidance, documentation, and reviewer-gated Vally evaluation coverage

## Validation

- `npm run check`
- 414 tests passed, 5 skipped
- all 11 new Python bad-app scenarios completed through the generated Tier 1 `run-all.ps1`
- `npx vally lint --eval-spec evals/azure-functions-doctor/eval.yaml --strict --verbose`

LLM-backed Vally execution remains restricted to the reviewer-gated GitHub Environment.

Closes #203

### CODE REVIEW

File: evals/azure-functions-doctor/fixtures/python-semantic/function_app.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import azure.functions as func
+import requests
+
+app = func.FunctionApp()
+jobs = func.Blueprint()
+
+
+@jobs.queue_trigger(
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/function_app.py]: This will work but its not the correct implementation for blueprints. Use this example for reference: https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-python?pivots=python-mode-decorators#organizing-with-blueprints

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/host.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (gavin-aguiar) [inline tests/fixtures/doctor-bad-apps/python-deploy-artifacts/host.json:1]: Does host.json require some more configs to be set for durable apps?
https://learn.microsoft.com/en-us/azure/durable-task/durable-functions/quickstart-python-vscode#configuration

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+azure-functions==1.21.0
+azure-functions-worker==1.0.0
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Remove `azure-functions-worker`

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+azure-functions==1.21.0
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Do not pin `azure-functions` . Let it use the latest

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+azure-functions-worker==1.0.0
+numpy==2.0.0
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Remove the version from all the requirements.

File: evals/azure-functions-doctor/fixtures/python-semantic/function_app.py (hunk 1 above)
Comment (gavin-aguiar) [reply evals/azure-functions-doctor/fixtures/python-semantic/function_app.py]: Update all the other function apps too

**Review by gavin-aguiar — CHANGES_REQUESTED** [review]

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/requirements.lock [hunk 6]
```diff
@@ -0,0 +1 @@
+azure-functions==1.21.0
```
Comment (gavin-aguiar) [inline tests/fixtures/doctor-bad-apps/python-deploy-artifacts/requirements.lock]: we dont need this file

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/tests/test_app.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+def test_placeholder() -> None:
```
Comment (gavin-aguiar) [inline tests/fixtures/doctor-bad-apps/python-deploy-artifacts/tests/test_app.py]: We can remove this test file too

File: tests/doctor-context.test.ts [hunk 8]
```diff
@@ -49,6 +49,30 @@ describe('loadProjectContext', () => {
...
+    const ctx = await loadProjectContext(dir);
+
+    expect(ctx.language).toBe('python');
+    expect(ctx.python?.programmingModel).toBe('v2');
+  });
+
+  it('uses the explicit worker runtime when package.json is only project tooling', async () => {
+    const dir = makeTmp('doctor-ctx-python-worker-');
+    writeFileSync(join(dir, 'host.json'), JSON.stringify({ version: '2.0' }));
+    writeFileSync(join(dir, 'package.json'), JSON.stringify({ private: true }));
```
Comment (gavin-aguiar) [inline tests/doctor-context.test.ts]: Why is package.json added in a python app?

**Review by gavin-aguiar — CHANGES_REQUESTED** [review]

File: evals/azure-functions-doctor/fixtures/python-semantic/function_app.py (hunk 1 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/function_app.py]: Updated the Blueprint example to follow the documented multi-file organization: the Blueprint now lives in `jobs.py` and `function_app.py` imports it. Registration remains intentionally absent because this semantic fixture must exercise the unregistered-Blueprint diagnostic.

Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/function_app.py]: Updated the corresponding bad-app fixture as well: its Blueprint now lives in a separate `jobs.py`, while `function_app.py` intentionally omits `app.register_functions(jobs)` so the expected diagnostic still fires.

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/host.json (hunk 2 above)
Comment (TsuyoshiUshio) [reply tests/fixtures/doctor-bad-apps/python-deploy-artifacts/host.json:1]: This fixture is not a Durable Functions app; it is an HTTP app used only to detect deployable test/cache artifacts and a missing dependency lockfile, so Durable host settings do not apply. The separate `python-durable-defaults` fixture intentionally omits `extensions.durableTask` to exercise that warning, and the expected-results documentation now makes that intent explicit.

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt (hunk 3 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Removed `azure-functions-worker` from the semantic fixture requirements.

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt (hunk 4 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Removed the `azure-functions` version pin so the fixture uses the latest compatible package.

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt (hunk 5 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Removed versions from all requirements in the semantic fixture. The eval prompt now treats unpinned/unlocked dependencies as the deterministic Tier 1 finding.

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/requirements.lock (hunk 6 above)
Comment (TsuyoshiUshio) [reply tests/fixtures/doctor-bad-apps/python-deploy-artifacts/requirements.lock]: Removed `requirements.lock`. The fixture now correctly exercises the missing-lockfile diagnostic.

File: tests/fixtures/doctor-bad-apps/python-deploy-artifacts/tests/test_app.py (hunk 7 above)
Comment (TsuyoshiUshio) [reply tests/fixtures/doctor-bad-apps/python-deploy-artifacts/tests/test_app.py]: Removed the synthetic test file. A `.gitkeep` preserves the intentionally deployable `tests/` directory so the deploy-artifact diagnostic remains covered after cloning.

File: tests/doctor-context.test.ts (hunk 8 above)
Comment (TsuyoshiUshio) [reply tests/doctor-context.test.ts]: This test models a Python app in a repository that also has JavaScript tooling (for example, ESLint). I renamed the test and added realistic tooling-only scripts/devDependencies to make that explicit; `FUNCTIONS_WORKER_RUNTIME=python` must take precedence over the presence of `package.json`.

File: src/doctor/python-dependencies.ts [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    if (!match || normalizePackageName(match[1]) === 'python') continue;
+    const dependency = parseRequirement(
+      `${match[1]}${match[2]}`,
+      'pyproject.toml',
+      index + 1,
+    );
+    if (dependency && !manifest.dependencies.some(existing => existing.name === dependency.name)) {
+      manifest.dependencies.push(dependency);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/doctor/python-dependencies.ts]: Fixed Poetry parsing to iterate over the complete `pyproject.toml`, so reported locations are absolute file line numbers. Added a regression assertion that `orjson` is reported at line 10.

File: src/doctor/python-source.ts [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+      continue;
+    }
+    if (!entry.isFile() || !entry.name.endsWith('.py')) continue;
+    const relativePath = relative(root, path);
+    files.push({
+      path,
+      relativePath,
+      module: moduleName(relativePath),
+      source: readFileSync(path, 'utf-8'),
+    });
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/doctor/python-source.ts:85]: Added a shared safe Python reader and threaded an injectable reader through recursive v2 discovery. Unreadable or concurrently deleted `.py` files are skipped while readable functions are still discovered; a regression test covers this behavior.

File: src/doctor/python-dependencies.ts [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+export function hasExternalPythonImports(dir: string): boolean {
+  const sourceFiles = listPythonSourceFiles(dir);
+  const localModules = new Set(
+    sourceFiles.map(path => relative(dir, path).split(/[\\/]/)[0].replace(/\.py$/, '')),
+  );
+  for (const path of sourceFiles) {
+    const content = readFileSync(path, 'utf-8');
+    for (const line of content.split(/\r?\n/)) {
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/doctor/python-dependencies.ts:286]: Updated external-import scanning to use the same safe Python reader and continue past unreadable or concurrently deleted files. Added an injected-reader regression test for the non-fatal path.

File: evals/azure-functions-doctor/fixtures/python-semantic/function_app.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import azure.functions as func
+import requests
+from jobs import jobs
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/function_app.py:5]: Jobs needs to be registered in this file., 
`app.register_functions(jobs)`

File: evals/azure-functions-doctor/fixtures/python-semantic/jobs.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+jobs = func.Blueprint()
+
+
+@jobs.queue_trigger(
+    arg_name="message",
+    queue_name="jobs",
+    connection="AzureWebJobsStorage",
+)
+def process_job(message: func.QueueMessage) -> None:
+    print(message.get_body().decode())
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/jobs.py]: Use logging.info instead of print

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,3 @@
+azure-functions
+numpy
+requests
```
Comment (gavin-aguiar) [inline evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: These are not used.

File: evals/azure-functions-doctor/fixtures/python-semantic/function_app.py (hunk 12 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/function_app.py:5]: Registered the Blueprint with `app.register_functions(jobs)` in the semantic eval fixture. I also updated the Tier 1 context and grader so this fixture now expects Blueprint registration to pass. The separate bad-app fixture remains intentionally unregistered to test that diagnostic.

File: evals/azure-functions-doctor/fixtures/python-semantic/jobs.py (hunk 13 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/jobs.py]: Replaced `print` with `logging.info` and added the logging import.

File: evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt (hunk 14 above)
Comment (TsuyoshiUshio) [reply evals/azure-functions-doctor/fixtures/python-semantic/requirements.txt]: Removed both `numpy` and `requests`. The fixture now uses the standard-library `urllib.request.urlopen` to preserve the intentional synchronous HTTP call inside an async handler without an extra application dependency; the eval prompt and grader were updated accordingly.

**Review by gavin-aguiar — APPROVED** [review]
